lib/spack/llnl/util/filesystem.py

```diff
@@ -833,7 +848,24 @@ def names(self):
         Returns:
             list of strings: A list of library names
         """
-        return list(dedupe(x.split('.')[0][3:] for x in self.basenames))
+        names = []
+
+        for x in self.basenames:
+            name = x
+            if x.startswith('lib'):
+                name = x[3:]
+
+            # Valid extensions include: ['.dylib', '.so', '.a']
+            for ext in ['.dylib', '.so', '.a']:
+                i = name.rfind(ext)
+                if i != -1:
+                    names.append(name[:i])
+                    break
+            else:
+                # No valid extension, should we still include it?
```

Comment:
to clarify, `libmpi.so.20.10.1` can exist, but i guess one always have `libmpi.so` around symlinked to it.